airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/message/AirbyteType.kt

```diff
@@ -0,0 +1,157 @@
+package io.airbyte.cdk.message
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.fasterxml.jackson.databind.node.ObjectNode
+
+sealed class AirbyteType {
+    companion object {
+        fun fromAirbyteJsonSchema(jsonSchema: ObjectNode): ObjectType {
+            val airbyteSchema = fromJsonSchema(jsonSchema)
+            if (airbyteSchema !is ObjectType) {
+                throw IllegalArgumentException("Top-level schema must be an object (got $airbyteSchema)")
+            }
+            return airbyteSchema
+        }
+
+        fun fromJsonSchema(schema: ObjectNode): AirbyteType {
+            if (schema.has("type")) {
+                val schemaType = schema.get("type")
+                if (schemaType.isTextual) {
+                    return when (schema.get("type").asText()) {
+                        "string" -> StringType.fromJsonSchema(schema)
+                        "boolean" -> BooleanType
+                        "integer" -> IntegerType
+                        "number" -> NumberType.fromJsonSchema(schema)
+                        "array" -> ArrayType.fromJsonSchema(schema)
+                        "object" -> ObjectType.fromJsonSchema(schema)
+                        "null" -> NullType
+                        else -> throw IllegalArgumentException(
+                            "Unknown type: ${
+                                schema.get("type").asText()
+                            }"
+                        )
+                    }
+                } else if (schemaType.isArray) {
+                    return UnionType.fromJsonSchemaTypeArray(schemaType as ArrayNode, schema)
+                }
+            } else {
+                val options = schema.get("oneOf") ?: schema.get("anyOf") ?: schema.get("allOf")
+                if (options != null) {
+                    return UnionType.fromUnionOptions(
+                        options.map { fromJsonSchema(it as ObjectNode) }
+                    )
+                }
+            }
+
+            throw IllegalArgumentException("Unknown schema type: $schema")
+        }
+    }
+}
+
+data object NullType: AirbyteType()
```

Comment:
`{"type": "null"}` is [legal jsonschema](https://json-schema.org/understanding-json-schema/reference/null), and I've seen it in production schemas.

`Field(NullType, nullable = true)` == `{"type": null, "required": false}` and `Field(NullType, nullable = false)` == `{"type": "null", "required": true}` :(

Reply:
I do agree that we should force all the client data (except the top-level object) to be nullable.

Metadata I'm agnostic on. Currently it's non-nullable in the avro/parquet schemas for example. I don't know what you do in db schemas.

Reply:
blegh. didn't know that we explicitly use type:null. what does that even mean (i.e. why would a field always be null)? >.>

metadata is probably a weird mix in db/dw land. Some are nullable, purely for the sake of migrations (i.e. we added generation_id later, so probably did `alter table add generation_id null default null`). I wouldn't index too heavily on this

this is one of the reasons why I think treating the meta fields specially could make sense - they're the fields that typically fall into weird backwards-compat behavior. but I also think it's fine to have them here, since it'll make it easier for us to handle differences in that area in the future :shrug:  

I think there's a decent chance that eventually, destinations will have stuff like `if (schema.properties["airbyte_generation"].nullable)` somewhere in their deduping SQL queries. Which feels fine?